services/src/main/java/org/keycloak/vault/ElytronCSKeyStoreProviderFactory.java

```diff
@@ -0,0 +1,198 @@
+        this.credentialStoreLocation = config.get(CS_LOCATION);
+        if (this.credentialStoreLocation == null) {
+            logger.warn("ElytronCSKeyStoreProviderFactory not properly configured - missing store location");
+            return;
+        }
+        if (!Files.exists(Paths.get(this.credentialStoreLocation))) {
+            throw new VaultNotFoundException("The " + this.credentialStoreLocation + " file doesn't exist");
+        }
+
+        this.credentialStoreSecret = config.get(CS_SECRET);
+        if (this.credentialStoreSecret == null) {
+            logger.warn("ElytronCSKeyStoreProviderFactory not properly configured - missing store secret");
+            return;
+        }
+        this.credentialStoreType = config.get(CS_TYPE, "JCEKS");
+
+        // install the elytron credential store provider.
+        Security.addProvider(WildFlyElytronCredentialStoreProvider.getInstance());
+    }
+
+    @Override
+    public void postInit(KeycloakSessionFactory factory) {
+    }
+
+    @Override
+    public void close() {
+        // remove the elytron credential store provider.
+        Security.removeProvider(WildFlyElytronCredentialStoreProvider.getInstance().getName());
+    }
+
+    @Override
+    public String getId() {
+        return PROVIDER_ID;
+    }
+
+    /**
+     * Obtains the name of realm from the {@link KeycloakSession}.
+     *
+     * @param session a reference to the {@link KeycloakSession}.
+     * @return the name of the realm.
+     */
+    protected String getRealmName(KeycloakSession session) {
+        return session.getContext().getRealm().getName();
+    }
+
+    /**
+     * Obtains the {@code CredentialSource} to be used as a protection parameter when initializing the Elytron credential
+     * store. The source is essentially a wrapper for the credential store secret. The credential store secret can be specified
+     * in clear text form or in masked form. Check the Elytron tool documentation for instruction on how to mask the credential
+     * store secret.
+     *
+     * @param secret the secret obtained from the {@link ElytronCSKeyStoreProviderFactory} configuration.
+     * @return the constructed {@code CredentialSource}.
+     */
+    protected CredentialSource getCredentialSource(final String secret) {
+        if (secret != null && secret.startsWith("MASK-")) {
+            return new CredentialSource() {
+                @Override
+                public SupportLevel getCredentialAcquireSupport(Class<? extends Credential> credentialType, String algorithmName, AlgorithmParameterSpec parameterSpec) throws IOException {
+                    return credentialType == PasswordCredential.class ? SupportLevel.SUPPORTED : SupportLevel.UNSUPPORTED;
+                }
+
+                @Override
+                public <C extends Credential> C getCredential(Class<C> credentialType, String algorithmName, AlgorithmParameterSpec parameterSpec) throws IOException {
```

Comment:
This functions seems to me like something which shouldn't be implemented by us. Is there something provided by elytron which can be used? Because masked secret is provided by elytron right? So in case something is changed there, we would need to update this, which could be avoided in case we use something provided by them. WDYT?

Reply:
Yeah, but Elytron doesn't provide this out of the box. A similar function exists in the elytron subsystem in WildFly to read the secret strings for the credential stores. I agree it would be better to delegate this instead of doing it ourselves but that's just the way it is right now.

I'll ping Darran and ask if this could be provided by Elytron somehow. If they do it in the future we could replace this code for something that simply delegates to Elytron.